phoenix-core/src/it/java/org/apache/phoenix/end2end/index/ImmutableIndexExtendedIT.java

```diff
@@ -0,0 +1,258 @@
+            ResultSet rs = conn.createStatement().executeQuery(dql);
+            assertFalse(rs.next());
+
+            verifyRowCountForEmptyCol(conn, indexTable, initialRowCount, 0);
+
+            TestUtil.removeCoprocessor(conn, indexTable, PreMutationFailingRegionObserver.class);
+        }
+    }
+
+    @Test
+    public void testGlobalImmutableIndexUpsertPostMutationFailure() throws Exception {
+        String dataTable = "TBL_" + generateUniqueName();
+        String indexTable = "IND_" + generateUniqueName();
+
+        try (Connection conn = DriverManager.getConnection(getUrl())) {
+            final int initialRowCount = 2;
+            createAndPopulateTable(conn, dataTable, initialRowCount);
+            createIndex(conn, dataTable, indexTable);
+            TestUtil.addCoprocessor(conn, indexTable, PostMutationFailingRegionObserver.class);
+
+            boolean upsertfailed = false;
+            try {
+                conn.createStatement().execute("UPSERT INTO " + dataTable +
+                        " VALUES (3, 'a3', 'ab3', 'abc3')");
+                System.out.println("upsert initiated");
+                conn.commit();
+            } catch (Exception ex)
+            {
+                upsertfailed = true;
+            }
+            assertEquals(true, upsertfailed);
+
+            // verify that the row was not inserted into the data table
+            String dql = "SELECT * FROM " + dataTable + " WHERE id = 3";
+            ResultSet rs = conn.createStatement().executeQuery(dql);
+            assertFalse(rs.next());
+
+            verifyRowCountForEmptyCol(conn, indexTable, initialRowCount, 1);
+
+            TestUtil.removeCoprocessor(conn, indexTable, PostMutationFailingRegionObserver.class);
+        }
+    }
+
+    @Test
+    public void testGlobalImmutableIndexUpsertRetry() throws Exception {
+        String dataTable = "TBL_" + generateUniqueName();
+        String indexTable = "IND_" + generateUniqueName();
+
+        try (Connection conn = DriverManager.getConnection(getUrl())) {
+            final int initialRowCount = 2;
+            createAndPopulateTable(conn, dataTable, initialRowCount);
+            createIndex(conn, dataTable, indexTable);
+            TestUtil.addCoprocessor(conn, indexTable, FailOnceMutationRegionObserver.class);
+
+            conn.createStatement()
+                    .execute("UPSERT INTO " + dataTable + " VALUES (3, 'a3', 'ab3', 'abc3')");
+            conn.commit();
+
+            // verify that the row was inserted into the data table
+            String dql = "SELECT * FROM " + dataTable + " WHERE id = 3";
+            ResultSet rs = conn.createStatement().executeQuery(dql);
+            assertTrue(rs.next());
+            assertEquals(3, rs.getInt(1));
+
+            verifyRowCountForEmptyCol(conn, indexTable, initialRowCount + 1, 0);
+
+            TestUtil.removeCoprocessor(conn, indexTable, FailOnceMutationRegionObserver.class);
+        }
```

Comment:
These tests look similar with the difference of Coprocessor class. Can you refactor them and just call a function with Coprocessor class?